### PR TITLE
Upgrading schema compiler to 13.5.1

### PR DESCRIPTION
**Pull request by achannn**

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -73,7 +73,13 @@ for WORKER in "${WORKER_DIRS[@]}"; do
   # Compile the schema
   OUT_DIR=improbable/generated
   mkdir -p "$OUT_DIR"
-  "$DOWNLOAD_DIR"/schema_compiler --schema_path="${BUILD_DIR}/SpatialOS/schema" --schema_path="$DOWNLOAD_DIR" --csharp_out="$OUT_DIR" --load_all_schema_on_schema_path "${BUILD_DIR}"/SpatialOS/schema/*.schema "${DOWNLOAD_DIR}"/improbable/*.schema
+  "$DOWNLOAD_DIR"/schema_compiler \
```
Comment (zeroZshadow) [inline build.sh]: Instead of generating the same csharp in a loop for every worker, this could be combined with the schema description generator.
Generate both the descriptor and csharp in 1 go, then copy the csharp to each worker.

File: build.sh [hunk 2]
```diff
@@ -91,15 +97,14 @@ for WORKER in "${WORKER_DIRS[@]}"; do
...
-"${DOWNLOAD_DIR}/schema_compiler" --schema_path="${BUILD_DIR}/SpatialOS/schema" --schema_path="${DOWNLOAD_DIR}" --proto_out="${DOWNLOAD_DIR}/generated_protos" --load_all_schema_on_schema_path "${DOWNLOAD_DIR}"/improbable/*.schema "${BUILD_DIR}"/SpatialOS/schema/*.schema
-"${DOWNLOAD_DIR}/protoc" --proto_path="${DOWNLOAD_DIR}/generated_protos" --descriptor_set_out="${BUILD_DIR}/SpatialOS/schema/bin/schema.descriptor" --include_imports "${DOWNLOAD_DIR}"/generated_protos/*.proto "${DOWNLOAD_DIR}"/generated_protos/**/*.proto
-
-rm -r "${DOWNLOAD_DIR}/generated_protos"
+"${DOWNLOAD_DIR}/schema_compiler" \
+  --schema_path="${BUILD_DIR}/SpatialOS/schema" \
+  --schema_path="${DOWNLOAD_DIR}" \
+  --load_all_schema_on_schema_path \
+  --descriptor_set_out="${BUILD_DIR}/SpatialOS/schema/bin/schema.descriptor" \
+  ${DOWNLOAD_DIR}/improbable/*.schema \
```
Comment (zeroZshadow) [inline build.sh]: Just wondering, why is this not in quotes while others are? (I'm not experienced in bash, there might be a reason for this?)

Comment (achannn) [reply build.sh]: Have changed it to make it more consistent.

File: build.sh (hunk 1 above)
Comment (achannn) [reply build.sh]: No longer generating the csharp in a loop.  However, have still kept the csharp and schema descriptor generation in two discrete steps to make it seem less magical.

File: build.sh [hunk 3]
```diff
@@ -37,51 +37,58 @@ function getPlatformName() {
...
+  if [ ! -d "${TARGETDIR}" ]; then
+    spatial package get --force --unzip "${TYPE}" "${PACKAGE}" "${SDK_VERSION}" "${TARGETDIR}"
   fi
   popd
 }
 
 # Get the tools:
 # * Spatial Schema compiler
 # * Standard Library Schemas
-# * Core SDK for all platforms to enable building workers for MacOS, Windows or Linux
```
Comment (JonasImprobable) [inline build.sh]: Why was this line removed?

File: build.sh [hunk 4]
```diff
@@ -97,14 +98,14 @@ for WORKER in "${WORKER_DIRS[@]}"; do
...
 done
 
-# Compile the schemas into a schema descriptor
-mkdir -p "${BUILD_DIR}/SpatialOS/schema/bin"
-"${DOWNLOAD_DIR}/schema_compiler" \
-  --schema_path="${BUILD_DIR}/SpatialOS/schema" \
-  --schema_path="${DOWNLOAD_DIR}" \
+# Generate a schema descriptor from the schemas:
+mkdir -p "${BUILD_DIR}"/SpatialOS/schema/bin
+"${DOWNLOAD_DIR}"/schema_compiler/schema_compiler \
```
Comment (zeroZshadow) [inline build.sh]: This can be combined with the c# generation on line 71

File: build.sh [hunk 5]
```diff
@@ -37,51 +37,58 @@ function getPlatformName() {
...
+# * Core SDK for all platforms to enable building workers for MacOS, Windows or Linux
+retrievePackage "tools" "schema_compiler-x86_64-${PLATFORM_NAME}" "schema_compiler"
+retrievePackage "schema" "standard_library" "standard_library"
+retrievePackage "worker_sdk" "csharp" "lib/csharp"
+retrievePackage "worker_sdk" "core-dynamic-x86_64-win32" "lib/win64"
+retrievePackage "worker_sdk" "core-dynamic-x86_64-linux" "lib/linux64"
+retrievePackage "worker_sdk" "core-dynamic-x86_64-macos" "lib/macos64"
+
+# Generate C# code from the schemas:
+mkdir -p "${DOWNLOAD_DIR}"/csharp
```
Comment (JonasImprobable) [inline build.sh]: Nit: How about creating a OUT_DIR variable that we can reuse in line 74? That would make it more clear what we are doing here. Also, how about renaming the dir name slightly since (1) conceptually, it's not part of the our worker packages, how about putting the folder into something like`/SpatialOS/schema/generated/csharp`? While we are at it, how about factoring `${BUILD_DIR}"/SpatialOS/schema` into its own variable (something like `SCHEMA_DIR`)?

File: build.sh (hunk 4 above)
Comment (JonasImprobable) [reply build.sh]: My personal opinion on this would be that the current state is fine. Both steps can technically be achieved by a single schema_compiler invocation but having two clearly separated steps might make this easier to understand for users i.e. there is one less gotcha when you try to understand the code. What do you think?

File: build.sh [hunk 6]
```diff
@@ -37,51 +37,58 @@ function getPlatformName() {
...
-  cp "$DOWNLOAD_DIR"/*.so improbable/dependencies/native
+  cp -R "${DOWNLOAD_DIR}"/csharp/* improbable/generated
+  cp "${DOWNLOAD_DIR}"/lib/csharp/* improbable/dependencies/managed
+  cp "${DOWNLOAD_DIR}"/lib/win64/* improbable/dependencies/native
+  cp "${DOWNLOAD_DIR}"/lib/linux64/* improbable/dependencies/native
+  cp "${DOWNLOAD_DIR}"/lib/macos64/* improbable/dependencies/native
+
   for PLATFORM in "${BUILD_PLATFORMS[@]}"; do
     ${BUILD_TOOL} CsharpWorker.sln /property:Configuration=Release /property:Platform="$PLATFORM"
     cp -r bin ..
```
[the remark this answers is not in the document]
Comment (achannn) [reply build.sh]: We would have to prepend src to all the other paths then; not sure if it's worth removing this.

[the remark this answers is not in the document]
Comment (JonasImprobable) [reply build.sh]: Yeah, I guess the more interesting question should be why the .sln and .csproj files were put into the src folder in the first place. I'm ok with this for now I guess.

File: build.sh [hunk 7]
```diff
@@ -6,9 +6,9 @@ cd "$(dirname "$0")"
 
 WORKER_DIRS=(HelloWorker OtherWorkers/DiceWorker OtherWorkers/Interactive/client)
 BUILD_PLATFORMS=(macOS64 Windows64 Linux64)
-DOWNLOAD_DIR="$(pwd)/build"
+DOWNLOAD_DIR="$(pwd)/packages"
```
Comment (JonasImprobable) [inline build.sh]: Super nit: 
```suggestion
PACKAGES_DIR="$(pwd)/packages"
```
